power_grid_model_c/power_grid_model_cpp/include/power_grid_model_cpp/handle.hpp

```diff
@@ -0,0 +1,86 @@
+// SPDX-FileCopyrightText: Contributors to the Power Grid Model project <powergridmodel@lfenergy.org>
+//
+// SPDX-License-Identifier: MPL-2.0
+
+#pragma once
+#ifndef POWER_GRID_MODEL_CPP_HANDLE_HPP
+#define POWER_GRID_MODEL_CPP_HANDLE_HPP
+
+#include "power_grid_model_c/handle.h"
+
+#include "basics.hpp"
+
+namespace power_grid_model_cpp {
+class PowerGridError : public std::exception {
+  public:
+    PowerGridError(const std::string& message) : message_(message) {}
+    const char* what() const noexcept override { return message_.c_str(); }
+
+  private:
+    std::string message_;
+};
+
+class PowerGridRegularError : public PowerGridError {
+  public:
+    using PowerGridError::PowerGridError;
+    static constexpr Idx error_code() { return PGM_regular_error; }
+};
+
+class PowerGridSerializationError : public PowerGridError {
+  public:
+    using PowerGridError::PowerGridError;
+    static constexpr Idx error_code() { return PGM_serialization_error; }
+};
+
+class PowerGridBatchError : public PowerGridError {
+  public:
+    struct FailedScenario {
+        Idx scenario;
+        std::string error_message;
+    };
+
+    PowerGridBatchError(std::string const& message, std::vector<FailedScenario> failed_scenarios_c)
+        : PowerGridError{message}, failed_scenarios_{std::move(failed_scenarios_c)} {}
+    static constexpr Idx error_code() { return PGM_batch_error; }
+    std::vector<FailedScenario> const& failed_scenarios() const { return failed_scenarios_; }
+
+  private:
+    std::vector<FailedScenario> failed_scenarios_;
+};
+
+class Handle {
+  public:
+    Handle() : handle_{PGM_create_handle()} {}
+
+    RawHandle* get() const { return handle_.get(); }
```

Comment:
@mgovers I see your other comments using `call_with`. Maybe this approach is cleaner from the user perspective?

Reply:
While I like the idea of adding a `ProxyHandle`, and I do think that from a user perspective, it looks quite nice, I think there are 2 arguments against it:
* It introduces a lot of intricate code that I believe is not necessary
* It requires a `throw`ing destructor, which is disallowed (and actually results in `std::terminate()`)

Reply:
Hmmm, I have not thought about the `throw` in destructor is an issue, haha. I think if the destructor is not called in the stack-unwinding process, it can technically throw. But you are correct.

Reply:
After some investigation it confirms my assumption. You can throw in the destructor if the destruction has not happend in the stack-unwinding process. In the destructor of `ProxyHandle` you can call `std::uncaught_exception` to check that condition.

Reply:
I do, however, feel like that's going to become very over-engineered. I would rather not dive into these kind of rare edge cases of the C++ specification.

If the user wants that kind of functionality and provide their own calls, IMO, they should just directly call the C API, create their own wrapper function, or be satisfied with just using `call_with` (name pending).

I do not believe the advantages of implementing the `ProxyHandle` outweigh the downsides.

Reply:
I have implemented it with the `with_call` function.